### PR TITLE
Allow to use constant adaptive index granularity for whole part

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimized memory usage for values of index granularity if granularity is constant for part. Added an ability to always select constant granularity for part (setting `use_const_adaptive_granularity`), which helps to ensure that it is always optimized in memory. It helps in large workloads (trillions of rows in shared storage) to avoid constantly growing memory usage by metadata (values of index granularity) of data parts.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit c5817d528c63c0a28bf373cbd46e072dcb8e254b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71786/c5817d528c63c0a28bf373cbd46e072dcb8e254b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71779/621cb60446cb17f0366f49b86c3432eed5db3716/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+{
+    size_t rows_before_mark = getMarkStartingRow(from_mark);
+    size_t last_row_pos = rows_before_mark + offset_in_rows + number_of_rows;
+    auto it = std::upper_bound(marks_rows_partial_sums.begin(), marks_rows_partial_sums.end(), last_row_pos);
+    size_t to_mark = it - marks_rows_partial_sums.begin();
+
+    return getRowsCountInRange(from_mark, std::max(1UL, to_mark)) - offset_in_rows;
+}
+
+void MergeTreeIndexGranularityAdaptive::shrinkToFitInMemory()
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.cpp:116]: maybe make it part of `optimize`?

File: src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#include <Storages/MergeTree/MergeTreeIndexGranularity.h>
+
+namespace DB
+{
+
+/// Class that stores adaptive index granularity.
+/// Inside it contains vector of partial sums of rows after mark:
+/// |-----|---|----|----|
+/// |  5  | 8 | 12 | 16 |
+class MergeTreeIndexGranularityAdaptive : public MergeTreeIndexGranularity
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.h:11]:
```suggestion
class MergeTreeIndexGranularityAdaptive final : public MergeTreeIndexGranularity
```

File: src/Storages/MergeTree/MergeTreeIndexGranularityConstant.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
+#pragma once
+#include <Storages/MergeTree/MergeTreeIndexGranularity.h>
+
+namespace DB
+{
+
+/// Class that stores constant index granularity for whole part, except
+/// last non-zero granule and final granule which always has zero rows.
+class MergeTreeIndexGranularityConstant : public MergeTreeIndexGranularity
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeIndexGranularityConstant.h:9]:
```suggestion
class MergeTreeIndexGranularityConstant final : public MergeTreeIndexGranularity
```

**Review by alesapin — APPROVED** [review]
LGTM

File: src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.cpp (hunk 1 above)
Comment (CurtizJ) [reply src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.cpp]: Yes, makes sense

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
Maybe it also makes sense to randomize setting in CI

**Comment by CurtizJ** [comment]
I wanted to randomize settings in a new PR, but we can do it here.

**Comment by CurtizJ** [comment]
Random settings found a bug in #71990, so let's enable them in this PR.

**Comment by alesapin** [comment]
Completely green!

**Comment by CurtizJ** [comment]
Thanks, updated changelog.

**Comment by tanner-bruce** [comment]
I saw this in the changelog, we have a trillion row table, is there some indicators of when someone should look at using this?

**Comment by CurtizJ** [comment]
@tanner-bruce

In version 24.11 you run query `SELECT sum(index_granularity_bytes_in_memory_allocated) FROM system.parts` to check how much memory is consumed by internal structure optimized in this PR.

In general the indicator to use this feature is a noticeable constantly growing of memory usage.

**Pull request merged**